### PR TITLE
Group results

### PR DESCRIPTION
**Pull request by dealingwith**
Screenshot from 2024-02-13 21-45-27
Screenshot from 2024-02-13 22-43-03

### CODE REVIEW

File: suggest.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def image(link, title, content)
 end
 
 def strip_html_tags(html)
-  html.gsub(/<[^>]*>/, '')
+  html.gsub(/<\/?[^>]*>/, '')
```
Comment (dealingwith) [inline suggest.rb:17]: Better HTML stripping

File: suggest.rb [hunk 2]
```diff
@@ -52,28 +53,26 @@ def search_string_in_files(search_string)
   paragraph_count = 0
 
   Dir.foreach("./search") do |file_name|
-    next if file_name == '.' || file_name == '..'
+    next if file_name == '.' || file_name == '..' || file_name == original_file_name
```
Comment (dealingwith) [inline suggest.rb:56]: Don't include the original file in the search results. This shouldn't normally be an issue but it was when I was testing :)

File: suggest.rb [hunk 3]
```diff
@@ -52,28 +53,26 @@ def search_string_in_files(search_string)
...
 
     file_path = File.join("./search", file_name)
 
     if File.file?(file_path)
       file_content = sanitize_file_content(File.read(file_path))
       paragraphs = file_content.split(/\n/)
-      paragraphs.select! { |p| p.include?(search_string) }
+      paragraphs.select! { |p| p.match(/[^A-Za-z]+#{search_string}[^A-Za-z]+/i) }
```
Comment (dealingwith) [inline suggest.rb:63]: Improved to find results buried in URLs or with special characters before or after, for example an em dash

**Comment by dealingwith** [comment]
- [ ] Squash merge

File: suggest.rb [hunk 4]
```diff
@@ -138,21 +138,40 @@ def search_string_in_files(search_string)
 puts "searching for key words in searchable files"
 
 search_results = []
-chosen_key_words.each { |key_word| search_results << search_string_in_files(key_word) }
-
-ap search_results
+chosen_key_words.each { |key_word| search_results << search_string_in_files(key_word, filename) }
```
Comment (jamespeilunli) [inline suggest.rb:141]: You can use `Array#map` here if you want
```suggestion
search_results = chosen_key_words.map { |key_word| search_string_in_files(key_word, filename) }
```

**Review by jamespeilunli — APPROVED** [review]
LGTM (note that I can't test the code because I don't have an OpenAI key). I just added one small refactoring suggestion.

File: suggest.rb (hunk 4 above)
Comment (dealingwith) [reply suggest.rb]: Excellent, thanks!

**Comment by dealingwith** [comment]
> LGTM (note that I can't test the code because I don't have an OpenAI key). I just added one small refactoring suggestion.

I should have mentioned that these changes can all be tested using the cache, which is committed into the repo at the moment

**Comment by jamespeilunli** [comment]
@dealingwith this is the output I get when I try to run `ruby suggest.rb`:
```
  ____                                         _      ___    __  __           _     _        
 / ___|   _   _    __ _    __ _    ___   ___  | |_   / _ \  |  \/  |   __ _  | |_  (_)   ___ 
 \___ \  | | | |  / _` |  / _` |  / _ \ / __| | __| | | | | | |\/| |  / _` | | __| | |  / __|
  ___) | | |_| | | (_| | | (_| | |  __/ \__ \ | |_  | |_| | | |  | | | (_| | | |_  | | | (__ 
 |____/   \__,_|  \__, |  \__, |  \___| |___/  \__|  \___/  |_|  |_|  \__,_|  \__| |_|  \___|
                  |___/   |___/                                                              
/home/jamesli/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/ruby-openai-6.3.1/lib/openai.rb:61:in `access_token': OpenAI access token missing! See https://github.com/alexrudall/ruby-openai#usage (OpenAI::ConfigurationError)
        from /home/jamesli/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/ruby-openai-6.3.1/lib/openai/client.rb:20:in `block in initialize'
        from /home/jamesli/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/ruby-openai-6.3.1/lib/openai/client.rb:17:in `each'
        from /home/jamesli/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/ruby-openai-6.3.1/lib/openai/client.rb:17:in `initialize'
        from suggest.rb:85:in `new'
        from suggest.rb:85:in `<main>'
```
This is because the OpenAI initialization always runs no matter what:
line 85: `client = OpenAI::Client.new(access_token: ENV['OPENAI_API_KEY'])`

Should we only run the OpenAI setup if we don't use cache?

**Comment by dealingwith** [comment]
> This is because the OpenAI initialization always runs no matter what: line 85: `client = OpenAI::Client.new(access_token: ENV['OPENAI_API_KEY'])`
> 
> Should we only run the OpenAI setup if we don't use cache?

Ah! Yeah, I'll cut a fresh PR for that

**Comment by dealingwith** [comment]
> > This is because the OpenAI initialization always runs no matter what: line 85: `client = OpenAI::Client.new(access_token: ENV['OPENAI_API_KEY'])`
> > Should we only run the OpenAI setup if we don't use cache?
> 
> Ah! Yeah, I'll cut a fresh PR for that

https://github.com/dealingwith/som-ruby-cli-poc/pull/2/files#diff-d5e8b386f6ec484d952fbb306a3f1787557a553dcdb487a0e8cfc62fafad3c6b

**Pull request merged**